### DISCUSSION TITLE
test assertion failed fs_read_bufs due to faulty preadv() implementation

### DISCUSSION THREAD

**Discussion by iandjt** [libuv/libuv#4478] [General]
On Solaris 11.4 with libuv 1.48.0 I'm seeing this in 'make check' result:

```
not ok 96 - fs_read_bufs
# exit code 134

# Output from process `fs_read_bufs`:
# Assertion failed in test/test-fs.c on line 3212: `446 == uv_fs_read(((void *)0), &read_req, open_req1.result, bufs + 0, 2, 0, ((void *)0))` (446 == 256)

```
This looks to be because the preadv() emulation in unix/fs.c is faulty and is only using the first iov in a pread() ,
ignoring nbufs.

```
#define preadv(fd, bufs, nbufs, off)                                          \
  pread(fd, (bufs)->iov_base, (bufs)->iov_len, off)  

```

Looking at libuv 1.44.2 code, there seems to be a much better emulation of preadv() that used all the iov's via several pread()'s    

Perhaps a variant of that code should be reinstated.

I've attached a patch that works for me.

[fs.c_patch.txt](https://github.com/user-attachments/files/16387750/fs.c_patch.txt)

**Comment by santigimeno** [comment]
Have you checked with latest v1.x? It should have been addressed by https://github.com/libuv/libuv/commit/cc23e204d76688053c63fd1698435bc4b68aab6a

**Comment by iandjt** [comment]
Hi Santiago,
Sorry I was using the latest release, 1.48 which was in Feb 2024.  I see that patch you mention was after
that so maybe it will end up in 1.49

Thanks

Ian D
________________________________

**Comment by batrla** [comment]
@iandjt which Solaris release do you use to compile libuv? Note that Solaris 11.4 since SRU69 implements preadv and pwritev in the kernel. See: https://blogs.oracle.com/solaris/post/whats-new-in-oracle-solaris-114-sru69

The autodetection that was implemented in libuv as part of https://github.com/libuv/libuv/commit/cc23e204d76688053c63fd1698435bc4b68aab6a should hopefully work fine on older and newer Solaris releases, since the dlsym() should just find the real implementation on 11.4 SRU69+, otherwise libuv will just do the emulation on older Solaris releases.

```
$ nm /lib/amd64/libc.so.1 | egrep preadv\|pwritev | grep GLOB
[12843] |             2395040|                 298|FUNC |GLOB |3    |25     |preadv
[13747] |             2395680|                 298|FUNC |GLOB |3    |25     |pwritev

```

**Comment by iandjt** [comment]
Hi @batria, thanks for the info.  Alas our Solaris 11 is a bit older so we didn't have that new feature.

**Comment by saghul** [comment]
I don't think we support such an old Solaris version: https://github.com/libuv/libuv/blob/v1.x/SUPPORTED_PLATFORMS.md (should 121 be 12.1 though?)
